sway/desktop/launcher.c

```diff
@@ -79,45 +82,83 @@ struct sway_workspace *root_workspace_for_pid(pid_t pid) {
 			if (pid == _pw->pid) {
 				pw = _pw;
 				sway_log(SWAY_DEBUG,
-						"found pid_workspace for pid %d, workspace %s",
-						pid, pw->workspace);
-				goto found;
+					"found %s match for pid %d: %s",
+					node_type_to_str(pw->node->type), pid, node_get_name(pw->node));
+				break;
 			}
 		}
 		pid = get_parent_pid(pid);
 	} while (pid > 1);
 
-found:
-	if (pw && pw->workspace) {
-		ws = workspace_by_name(pw->workspace);
-
-		if (!ws) {
-			sway_log(SWAY_DEBUG,
-					"Creating workspace %s for pid %d because it disappeared",
-					pw->workspace, pid);
-
-			struct sway_output *output = pw->output;
-			if (pw->output && !pw->output->enabled) {
+	if (pw) {
+		switch (pw->node->type) {
+		case N_CONTAINER:
+			// Unimplemented
+			// TODO: add container matching?
+			ws = pw->node->sway_container->pending.workspace;
+			break;
+		case N_WORKSPACE:
+			ws = pw->node->sway_workspace;
+			break;
+		case N_OUTPUT:
+			output = pw->node->sway_output;
+			ws = workspace_by_name(pw->name);
+			if (!ws) {
 				sway_log(SWAY_DEBUG,
-						"Workspace output %s is disabled, trying another one",
-						pw->output->wlr_output->name);
-				output = NULL;
+						"Creating workspace %s for pid %d because it disappeared",
+						pw->name, pid);
+				if (!output->enabled) {
+					sway_log(SWAY_DEBUG,
+							"Workspace output %s is disabled, trying another one",
+							output->wlr_output->name);
+					output = NULL;
+				}
+				ws = workspace_create(output, pw->name);
 			}
-
-			ws = workspace_create(output, pw->workspace);
+			break;
+		case N_ROOT:
+			ws = workspace_create(NULL, pw->name);
+			break;
 		}
-
 		pid_workspace_destroy(pw);
 	}
 
 	return ws;
 }
 
-static void pw_handle_output_destroy(struct wl_listener *listener, void *data) {
-	struct pid_workspace *pw = wl_container_of(listener, pw, output_destroy);
-	pw->output = NULL;
-	wl_list_remove(&pw->output_destroy.link);
-	wl_list_init(&pw->output_destroy.link);
+static void pw_handle_node_destroy(struct wl_listener *listener, void *data) {
+	struct pid_workspace *pw = wl_container_of(listener, pw, node_destroy);
+	switch (pw->node->type) {
+	case N_CONTAINER:
+		// Unimplemented
+		break;
+	case N_WORKSPACE:;
+		struct sway_workspace *ws = pw->node->sway_workspace;
+		wl_list_remove(&pw->node_destroy.link);
+		wl_list_init(&pw->node_destroy.link);
```

Comment:
Can we move the `wl_list_remove()` + `wl_list_init()` out of the `switch`? Would make the logic a bit simpler.

Reply:
Yeah seems good.

Reply:
Wasn't sure a neat way to write this since I want to re-init the listener before I add it to another node_destroy signal, like at the end of this N_WORKSPACE branch.